src/main/java/ch/njol/skript/lang/function/DefaultFunction.java

```diff
@@ -0,0 +1,311 @@
+
+		private Contract contract = null;
+
+		private String[] description, since, examples, keywords;
+
+		private Builder(@NotNull String name, @NotNull Class<T> returnType) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(returnType, "return type cannot be null");
+
+			this.name = name;
+			this.returnType = returnType;
+		}
+
+		public Builder<T> contract(@NotNull Contract contract) {
+			Preconditions.checkNotNull(contract, "contract cannot be null");
+
+			this.contract = contract;
+			return this;
+		}
+
+		/**
+		 * Sets this function builder's description.
+		 *
+		 * @return This builder.
+		 */
+		public Builder<T> description(@NotNull String... description) {
+			Preconditions.checkNotNull(description, "description cannot be null");
+
+			this.description = description;
+			return this;
+		}
+
+		/**
+		 * Sets this function builder's version history.
+		 *
+		 * @return This builder.
+		 */
+		public Builder<T> since(@NotNull String... since) {
+			Preconditions.checkNotNull(since, "since cannot be null");
+
+			this.since = since;
+			return this;
+		}
+
+		/**
+		 * Sets this function builder's examples.
+		 *
+		 * @return This builder.
+		 */
+		public Builder<T> examples(@NotNull String... examples) {
+			Preconditions.checkNotNull(examples, "examples cannot be null");
+
+			this.examples = examples;
+			return this;
+		}
+
+		/**
+		 * Sets this function builder's keywords.
+		 *
+		 * @return This builder.
+		 */
+		public Builder<T> keywords(@NotNull String... keywords) {
+			Preconditions.checkNotNull(keywords, "keywords cannot be null");
+
+			this.keywords = keywords;
+			return this;
+		}
+
+		/**
+		 * Adds a parameter to this function builder.
+		 *
+		 * @param name The parameter name.
+		 * @param type The type of the parameter.
+		 * @return This builder.
+		 */
+		public <PT> Builder<T> parameter(@NotNull String name, @NotNull Class<PT> type) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(type, "type cannot be null");
+
+			parameters.put(name, new Parameter<>(name, getClassInfo(type), !type.isArray(), null));
+			return this;
+		}
+
+		/**
+		 * Adds an optional parameter to this function builder.
+		 *
+		 * @param name The parameter name.
+		 * @param type The type of the parameter.
+		 * @return This builder.
+		 */
+		public <PT> Builder<T> optionalParameter(@NotNull String name, @NotNull Class<PT> type) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(type, "type cannot be null");
+
+			parameters.put(name, new Parameter<>(name, getClassInfo(type), !type.isArray(), true));
+			return this;
+		}
+
+		/**
+		 * Completes this builder with the code to execute on call of this function.
+		 *
+		 * @param execute The code to execute.
+		 * @return The final function.
+		 */
+		public DefaultFunction<T> build(Function<FunctionArguments, T> execute) {
```

Comment:
I might prefer this to be split into just like
```
.execute(<function>)
.build()
```
but i'd like to see what others think